sync-global-pullsecret/sync-global-pullsecret.go

```diff
@@ -0,0 +1,227 @@
+	var err error
+
+	c, err := cmdutil.GetClient()
+	if err != nil {
+		return fmt.Errorf("failed to create client: %w", err)
+	}
+
+	// Create a new watcher
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to create file watcher: %w", err)
+	}
+	defer watcher.Close()
+
+	// Start watching the file
+	if err := watcher.Add(o.kubeletConfigJsonPath); err != nil {
+		return fmt.Errorf("failed to add file to watcher: %w", err)
+	}
+
+	// Initial check and fix
+	if err := o.checkAndFixFile(ctx, c); err != nil {
+		log.Printf("Initial file check failed: %v", err)
+	}
+
+	// Periodic check ticker
+	ticker := time.NewTicker(o.checkInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case event := <-watcher.Events:
+			if event.Has(fsnotify.Write) {
+				log.Printf("File change detected: %s", event.String())
+				if err := o.checkAndFixFile(ctx, c); err != nil {
+					log.Printf("Failed to fix file: %v", err)
+				}
+			}
+		case err := <-watcher.Errors:
+			log.Printf("Watcher error: %v", err)
+		case <-ticker.C:
+			if err := o.checkAndFixFile(ctx, c); err != nil {
+				log.Printf("Periodic check failed: %v", err)
+			}
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+// checkAndFixFile reads the current file content and updates it if it differs from the desired content.
+// Have in mind the logic which do the merge of the pull secret is in the globalpullsecret package under the HCCO.
+func (o *syncGlobalPullSecretOptions) checkAndFixFile(ctx context.Context, c client.Client) error {
+	// TODO (jparrill):
+	// 	- Validate the Kubelet flags does not contain a different path for the pull secret.
+
+	// Read existing content if file exists
+	existingContent, err := os.ReadFile(o.kubeletConfigJsonPath)
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("failed to read existing file: %w", err)
+	}
+
+	// Get the global pull secret from the DataPlane
+	globalPullSecret := &corev1.Secret{}
+	if err := c.Get(ctx, client.ObjectKey{Namespace: defaultGlobalPullSecretNamespace, Name: o.globalPSSecretName}, globalPullSecret); err != nil {
+		return fmt.Errorf("failed to get global pull secret: %w", err)
+	}
+
+	// Get bytes of the global pull secret
+	globalPullSecretBytes := globalPullSecret.Data[corev1.DockerConfigJsonKey]
+
+	// If file content is different, write the desired content
+	if string(existingContent) != string(globalPullSecretBytes) {
+		// Save original content for potential rollback
+		originalContent := existingContent
+
+		// Write the new content
+		if err := os.WriteFile(o.kubeletConfigJsonPath, globalPullSecretBytes, 0600); err != nil {
+			return fmt.Errorf("failed to write file: %w", err)
+		}
+		log.Printf("Pull secret updated %s with desired content", o.kubeletConfigJsonPath)
+
+		// Attempt to restart Kubelet with retries
+		maxRetries := 3
+		var lastErr error
+		for attempt := 1; attempt <= maxRetries; attempt++ {
+			if err := signalKubeletToRestartProcess(); err != nil {
```

Comment:
Is this logic taken from mco? if not, why? if yes can include comments pointing to the original source code?

Reply:
Humm not at all, they're restarting/reloading directly the systemd unit:

- https://github.com/openshift/machine-config-operator/blob/f30fb1b4ce118c59bbdf9c555a59f137b3909967/pkg/daemon/update.go#L249C85-L249C98
- https://github.com/openshift/machine-config-operator/blob/main/pkg/daemon/update.go#L78-L88